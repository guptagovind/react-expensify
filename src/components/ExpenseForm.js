import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props){
    super(props);
    this.state ={
      description: props.expense?props.expense.description:'',
      createdAt: props.expense?moment(props.expense.createdAt):moment(),
      calenderFocus:false,
      amount: props.expense? (props.expense.amount/100).toString():'',
      note: props.expense?props.expense.note:'',
      error:''
    }
  }


  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      this.setState(() => ({amount}));
    }
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note}));
  };

  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState(() => ({
        createdAt
      }))
    }
  };

  onFocusChange = ({focused}) => {
    this.setState(() => ({
      calenderFocus:focused
    }))
  };

  submit =(e) => {
    e.preventDefault();
    if(!this.state.description || !this.state.amount){
      this.setState(()=>({
        error:'Please provide description and amount'
      }))
    }else{
      this.setState(()=>({
        error:''
      }));
      this.props.onSubmit({
        description: this.state.description,
        createdAt: this.state.createdAt.valueOf(),
        amount: parseFloat(this.state.amount, 10) * 100,
        note: this.state.note,
      });
    }
  };

  render() {
    return (
        <form className="form" onSubmit={this.submit}>
          {this.state.error && <p className="form__error">{this.state.error}</p>}
          <input
            type="text"
            placeholder="Description"
            autoFocus
            className="text-input"
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            className="text-input"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocus}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            className="textarea"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <div>
            <button className="button">Save Expense</button>
          </div>

        </form>
    );
  }
}
