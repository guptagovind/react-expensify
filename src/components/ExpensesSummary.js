import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import selector from '../selectors/expenses';
import expenseTotal from '../selectors/expense-total';
import numeral from 'numeral';

export class ExpensesSummary extends React.Component {
  render() {
    return (
      <div className='page-header'>
        <div className='content-container'>
          <h1 className='page-header__title'>Viewing <span>{this.props.expensesCount}</span> {this.props.expensesCount === 1 ? 'expense' : 'expenses'} totalling <span>{numeral(this.props.expensesTotal / 100).format('$0,0.00')}</span></h1>
          <div className="page-header__actions">
            <Link className="button"to="/create">Add Expenses</Link>
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = (state)=>({
    expensesCount:selector(state.expenses, state.filters).length,
    expensesTotal:expenseTotal(selector(state.expenses, state.filters))
  });


export default connect(mapStateToProps)(ExpensesSummary);
