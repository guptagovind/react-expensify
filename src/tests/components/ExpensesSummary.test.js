import React from 'react';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import {shallow} from 'enzyme';

test('should show one expense with totalling', ()=>{
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={1000}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should show multiple expenses with totalling', ()=>{
  const wrapper = shallow(<ExpensesSummary expensesCount={4} expensesTotal={5000}/>);
  expect(wrapper).toMatchSnapshot();
});
