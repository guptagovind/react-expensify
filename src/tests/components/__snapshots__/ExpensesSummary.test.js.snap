// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show multiple expenses with totalling 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        4
      </span>
       
      expenses
       totalling 
      <span>
        $50.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expenses
      </Link>
    </div>
  </div>
</div>
`;

exports[`should show one expense with totalling 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $10.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expenses
      </Link>
    </div>
  </div>
</div>
`;
